execution-engine/src/fs.rs

```diff
@@ -334,20 +338,36 @@ impl FileSystem {
     // Operations on the filesystem. Rust style implementation of WASI API
     ////////////////////////////////////////////////////////////////////////////
 
-    /// The stub implementation of `clock_res_get`. Return unsupported error `NoSys`.
+    /// Get clock resolution from platform services. Return error `NoSys` in case of failure.
     #[inline]
-    pub(crate) fn clock_res_get(&self, _clock_id: ClockId) -> FileSystemResult<Timestamp> {
-        Err(ErrNo::NoSys)
+    pub(crate) fn clock_res_get(&self, clock_id: ClockId) -> FileSystemResult<Timestamp> {
+        if !self.enable_clock {
+            return Err(ErrNo::NoSys);
+        }
+
+        let clock_id = u32::from(clock_id) as u8;
+        match getclockres(clock_id) {
+            result::Result::Success(resolution) => Ok(Timestamp::from_nanos(resolution)),
+            _ => Err(ErrNo::NoSys),
```

Comment:
Is `NoSys` the right error code here? You don't seem to be distinguishing between the clock being disabled and it failing for some other reason, which will almost certainly be the `clock_id` being invalid, I would guess, though I'm a bit confused about which platform(s) are involved here.

Reply:
What error code do you suggest? I can't find any appropriate WASI error code for that.

Reply:
Presumably this is the list of possibilities: https://github.com/WebAssembly/WASI/blob/main/phases/snapshot/docs.md#errno

Reply:
On Linux there's `ENOSYS` for the thing not being implemented, I think, and `EINVAL` for the `clock_id` being unrecognised, and anything else is arguably an internal error and could perhaps cause an immediate exit.

Reply:
We're converting to WASI. The clock function call is made from the WASM program, goes through the execution engine before reaching the the platform services in the trusted runtime.
There doesn't seem to be any WASI error code for generic or internal errors, that's a shame.

Reply:
What do you think about panicking rather than returning a generic or internal error for the application to deal with? I find it hard to imagine how the application could do anything useful if it gets a value that suggests that the runtime is buggy.

Reply:
So this is the list of options, is it? https://docs.rs/wasi-types/0.1.6/wasi_types/enum.ErrNo.html

Reply:
Or is it here? https://github.com/veracruz-project/wasi-types/blob/veracruz/src/lib.rs#L120

Reply:
It looks like `NoSys` and `Inval` are the two errors we should expect to happen.

Reply:
I agree that there should be separate error codes returned, here.

Reply:
We should definitely return `inval` for unsupported clocks according to https://github.com/WebAssembly/WASI/blob/main/phases/snapshot/docs.md#-clock_res_getid-clockid---resulttimestamp-errno